ratis-server/src/test/java/org/apache/ratis/server/impl/LeaderElectionTests.java

```diff
@@ -97,6 +100,79 @@ public void testBasicLeaderElection() throws Exception {
     cluster.shutdown();
   }
 
+  static class SleepCode implements CodeInjectionForTesting.Code {
+    private final long sleepMs;
+
+    SleepCode(long sleepMs) {
+      this.sleepMs = sleepMs;
+    }
+
+    @Override
+    public boolean execute(Object localId, Object remoteId, Object... args) {
+      try {
+        LOG.info("{}: Simulate RaftServer startup blocking", localId);
+        Thread.sleep(sleepMs);
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      }
+      return true;
+    }
+  }
+
+  @Test
+  public void testWaitServerReady() throws Exception {
+    final int sleepMs = 1000 + ThreadLocalRandom.current().nextInt(1000);
+    LOG.info("Running testWaitServerReady, sleep = {}ms", sleepMs);
+    CodeInjectionForTesting.put(RaftServerImpl.START_COMPLETE, new SleepCode(sleepMs));
+    final MiniRaftCluster cluster = newCluster(1);
+    final Timestamp startTime = Timestamp.currentTime();
+    cluster.start();
+    LOG.info("Cluster started at {}ms", startTime.elapsedTimeMs());
+    final RaftGroupId groupId = cluster.getGroupId();
+    final RaftServerImpl server = (RaftServerImpl) cluster.getServers().iterator().next().getDivision(groupId);
+    final boolean isRunning = server.isRunning();
+    LOG.info("{} isRunning at {}ms? {}", server.getId(), startTime.elapsedTimeMs(), isRunning);
+
+    // Leader will be elected if the server is ready
+    Assertions.assertNotNull(waitForLeader(cluster), "No leader is elected.");
+    final long elapsedMs = startTime.elapsedTimeMs();
+    // allow a small difference to tolerate system timer inaccuracy
+    Assertions.assertTrue(elapsedMs > sleepMs - 10, () -> "elapseMs = " + elapsedMs + " but sleepMs = " + sleepMs);
+    cluster.shutdown();
+    CodeInjectionForTesting.remove(RaftServerImpl.START_COMPLETE);
+  }
+
+  @Test
+  public void testAddServerForWaitReady() throws IOException, InterruptedException {
+    LOG.info("Running testAddServerForWaitReady");
+    // normal startup cluster with 3 server
+    final MiniRaftCluster cluster = newCluster(3);
+    cluster.start();
+    RaftTestUtil.waitForLeader(cluster);
+    try (RaftClient client = cluster.createClient()) {
+      for (int i = 0; i < 10; ++i) {
+        RaftClientReply reply = client.io().send(new RaftTestUtil.SimpleMessage("message_" + i));
+        Assertions.assertTrue(reply.isSuccess());
+      }
+      // add 3 new servers and wait longer time
+      CodeInjectionForTesting.put(RaftServerImpl.START_COMPLETE, new SleepCode(2000));
```

Comment:
Okay, I fine to keep this test.  Then, we need to add `CodeInjectionForTesting.remove` at the end.

Reply:
ok